### ISSUE TITLE
[multi-gpu][direct bindings] nvfd.dtensor_execute should propagate `mesh_dim_names` on the output DTensor's mesh correctly.

### ISSUE THREAD

**Issue by kshitij12345** [NVIDIA/Fuser#4728]
Repro
```python
# torchrun --local-ranks-filter=0 --nnodes 1 --nproc-per-node 2 test_nvfuser.py
import torch.nn as nn
import torch
from torch.distributed._tensor import DeviceMesh, Shard, distribute_tensor
from torch.distributed.tensor.placement_types import Placement, Shard, Replicate
import os
import torch.distributed as dist

import torch.distributed as dist
from collections.abc import Iterable
from torch.distributed.tensor import DTensor
from torch.distributed.tensor.placement_types import Placement, Shard, Replicate
from typing import Callable, cast, TypeAlias

import torch
from dataclasses import dataclass
from functools import partial
from torch.distributed.tensor import DTensor
from functools import lru_cache
from enum import auto, Enum

import nvfuser_direct as nvfd
from nvfuser_direct import FusionDefinition as DirectFusionDefinition
from nvfuser_direct import DataType as DirectDataType


LOCAL_RANK = int(os.environ["LOCAL_RANK"])

device = torch.device("cuda", LOCAL_RANK)
torch.cuda.set_device(device)
num_devices = 2
mesh = DeviceMesh("cuda", list(range(num_devices)), mesh_dim_names=("tp",))

hidden_size = 16

class FusionDefinitionWrapper:
    def __init__(self, define_fusion: Callable[[DirectFusionDefinition], None]):
        """Wraps a function that defines a fusion without `multidevice_schedule`."""
        # The complete FusionDefinition (w/ multidevice_scheduler) will have to
        # be created at "call" time according to the input DTensors.
        self._define_fusion = define_fusion

    def _create_fusion_definition(
        self, in_dtensors: Iterable[DTensor]
    ) -> DirectFusionDefinition:
        def multidevice_schedule(fd: DirectFusionDefinition, in_dtensors: list[DTensor]) -> None:
            for in_tv, in_dtensor in zip(fd.fusion.inputs(), in_dtensors):
                assert isinstance(in_dtensor, DTensor)
                # Set the device mesh.
                assert (
                    in_dtensor.device_mesh.ndim == 1
                ), "nvFuser's Python API only supports 1D meshes."
                mesh = nvfd.multidevice.DeviceMesh(in_dtensor.device_mesh.mesh.tolist())

                in_tv.set_device_mesh(mesh)

                assert len(in_dtensor.placements) == 1, "Expect a 1D mesh"

                # Split and parallelize.
                # When the mesh is multi-dimensional, iterate through the
                # placements in descending order of Placement.dim.
                placement: Placement = in_dtensor.placements[0]
                if placement.is_shard():
                    dim = cast(Shard, placement).dim
                    in_tv.split(dim, mesh.size, inner_split=False)
                    in_tv.axis(dim).parallelize(nvfd.ParallelType.mesh_x)
                    in_tv.set_allocation_domain(in_tv.get_loop_domain(), new_contiguity=True)

        with DirectFusionDefinition() as fd:
            self._define_fusion(fd)
            multidevice_schedule(fd, in_dtensors)
        return fd

    def __call__(self, in_dtensors: Iterable[DTensor]) -> list[DTensor]:
        fusion_def = self._create_fusion_definition(in_dtensors)
        out_dtensors = nvfd.dtensor_execute(fusion_def, in_dtensors)
        return out_dtensors

def define_linear_forward(fd: DirectFusionDefinition) -> None:
    inp = fd.define_tensor([hidden_size, hidden_size], contiguity=True, dtype=nvfd.DataType.BFloat16)
    weight = fd.define_tensor([hidden_size, hidden_size], contiguity=True, dtype=nvfd.DataType.BFloat16)
    out = fd.ops.linear(inp, weight)
    fd.add_output(out)

fd = FusionDefinitionWrapper(define_linear_forward)

weight = distribute_tensor(torch.randn(hidden_size, hidden_size, requires_grad=False, dtype=torch.bfloat16), mesh, [Shard(0),])
in_dtensor = distribute_tensor(torch.randn(hidden_size, hidden_size, requires_grad=False, dtype=torch.bfloat16), mesh, [Replicate(),])


expected = torch.nn.functional.linear(in_dtensor, weight)

result = fd([in_dtensor, weight])

# Error because `mesh_dim_names` is not propagated.
torch.testing.assert_close(result[0], expected)
```

This can be a problem because mesh with different `mesh_dim_names` are not equal and can lead to error in downstream ops due to mixing different mesh.

```python
# torchrun --local-ranks-filter=0 --nproc_per_node=2 test.py
from torch.distributed.tensor import DTensor
from torch.distributed.device_mesh import init_device_mesh
import torch
from torch.distributed.tensor.placement_types import Shard

m_0 = init_device_mesh("cuda", (2,), mesh_dim_names=("tp",))

i_0 = DTensor.from_local(torch.randn(2, 2), m_0, [Shard(0),])

m_1 = init_device_mesh("cuda", (2,))

i_1 = DTensor.from_local(torch.randn(2, 2), m_1, [Shard(0),])

# Could not run pointwise computation across different mesh: Found DeviceMesh('cuda', [0, 1]) and DeviceMesh('cuda', [0, 1], mesh_dim_names=('tp',))!
i_0 + i_1
```

**Comment by wujingyue** [comment]
nvFuser exposes ParallelTypes, e.g., DIDx and DIDy. A 2D mesh, for example, will have axis 0 corresponding to DIDy and axis 1 DIDx. IIUC, mesh_dim_names essentially gives ParallelTypes custom names. 

I'm not yet clear how PyTorch tries to propagate the names and resolve conflicts. For example, what if a torch.add between `[Shard(0)]` and `[Shard(1)]`? What if a ReduceScatter from `[d, d], [Shard(0)]` to `[d], [Shard(0)]`? Assuming 1D mesh for both cases, will the same name be kept for the mesh axis? 

If the propagation rules require nvFuser to participate, we can add this feature into nvFuser. Otherwise, I'm inclined to do this in a "wrapper", e.g., `dtensor_execute` in `nvfuser_direct/__init__.py`.

**Comment by kshitij12345** [comment]
Based on my understanding, PyTorch only supports an operation if the inputs have been distributed across the same device mesh (though their individual placement could be different). Also, the output will be distributed across the same device mesh.

So, I don't think how the individual inputs are distributed across the mesh will affect the device mesh of the output (it will affect how it is distributed across that mesh).

I think `dtensor_execute` can just reuse the mesh from the input for the output.

```python
mesh = in_dtensors[0].device_mesh
out_dtensors: list[DTensor] = []
for out_tensor, out_sharding in zip(out_tensors, out_shardings):
    placements: list[Placement] = []
    for parallel_type in [_C_DIRECT.ParallelType.mesh_x]:
        axis: int = out_sharding.axis_sharded_on(parallel_type)
        placements.append(Replicate() if axis == -1 else Shard(axis))
    out_dtensors.append(DTensor.from_local(out_tensor, mesh, placements))
```

**Comment by wujingyue** [comment]
> I think dtensor_execute can just reuse the mesh from the input for the output.

That should be fine for now so LGTM. 

I think it'll break in the future for pipeline parallelism. For example, at the end of a pipeline stage, we may want to copy a tensor from a GPU to its neighbor. You may see the source DTensor fully allocated on GPU 0 only and the destination DTensor fully allocated on GPU 1. I think their device meshes have to be different -- I just haven't seen another way. Have you looked into existing pipeline parallelism implementations in PyTorch?

**Comment by kevinstephano** [comment]
@wujingyue should this still be an issue?
